Convert course home page to Bootstrap

andy-armstrong opened the pull request:
## [LEARNER-1694](https://openedx.atlassian.net/browse/LEARNER-1694)

### Description

This PR represents the completed version of my hackathon project. It converts the course home page to optionally use Bootstrap based upon a Waffle flag.

### Acceptance Criteria

- [ ] Add a Waffle flag so that the old and new versions of the page can both be supported
- [ ] Convert the page to work with Bootstrap

### Sandbox
- [ ] https://andy-armstrong.sandbox.edx.org/courses/course-v1:AndyA+VERIFIED+1/course/

Note: staff users are Waffle flagged on, so can see the Bootstrap version of the pages. Non-staff get the non-Bootstrap version, which is critical to test with as when we merge that is what folks will get initially. 

### Testing
Front-end changes:
- [ ] Accessibility (Check for a11y violations, ensure keyboard accessible, screen reader testing as appropriate)
- [ ] i18n
- [ ] RTL

### Assign Github reviewers (as needed)
- [ ] @HarryRein 
- [ ] @marcotuts 
- [ ] @AlasdairSwan
 
### Post-review
- [ ] Rebase and squash commits

andy-armstrong commented:
@marcotuts @HarryRein @AlasdairSwan The styling has gone awry after rebasing. I'm working on it now and will update the sandbox.

andy-armstrong commented:
@marcotuts @HarryRein @AlasdairSwan I'm baffled by the two column layout as it is working on my devstack but not on the sandbox. I'll have to investigate further tomorrow.

andy-armstrong commented:
@marcotuts @HarryRein @AlasdairSwan I changed my two-column layout to not wrap, but to instead use breakpoints to turn it into a one-column layout. It seems to be working well now, so this is ready for review.

HarryRein commented on cms/static/sass/bootstrap/_variables.scss:
```diff
@@ -57,7 +57,7 @@ $footer-bg: $white !default;
 $font-regular: 400 !default;
 $font-semibold: 600 !default;
-$font-bold: 700 !default;
+$font-bold: 600 !default;
```
Should we make semibold 500 to keep them different?

andy-armstrong replied:
Good catch. I don't know how that change snuck in, as I'm not working on Studio. I'll revert it.

AlasdairSwan replied:
@andy-armstrong the CSS keyword ```bold``` is ```700```, why move our bold variable to ```600```? I fear it may cause confusion.

andy-armstrong replied:
Agreed @AlasdairSwan. I don't know how that change got in (it must have been during the heat of the hackathon) but I've reverted it now.

HarryRein commented on lms/static/sass/features/_course-upgrade-message.scss:
```diff
@@ -3,11 +3,10 @@
 // Expanded upgrade message
 .vc-message {
-  background: $blue-d1;
+  background: $lms-hero-color;
```
Not necessary here, but what do you think we should do with some of the more obscure colors? Should we keep them in theme-colors?

andy-armstrong replied:
Good question, and I don't have a good answer for you. Let's discuss.

HarryRein commented on lms/static/sass/features/_course-experience.scss:
```diff
@@ -17,11 +20,11 @@
   .message-content {
     @include margin(0, 0, $baseline, $baseline);
 
+    flex-grow: 1;
```
<img width="891" alt="screen shot 2017-10-06 at 2 10 04 pm" src="https://user-images.githubusercontent.com/3861052/31292118-2de9b720-aaa0-11e7-8809-5d36273fc5d8.png">

andy-armstrong replied:
Whoops! Let me look at that!

andy-armstrong replied:
Fixed

HarryRein commented on lms/static/sass/features/_course-experience.scss:
```diff
@@ -152,9 +156,6 @@
-  @include margin-left(0);
-  @include padding-left($baseline);
-
   .section-tools li:not(:first-child) {
```
Something quirky happened to the alignment of course tools and their icons 
<img width="171" alt="screen shot 2017-10-06 at 2 14 13 pm" src="https://user-images.githubusercontent.com/3861052/31292256-b6681902-aaa0-11e7-970e-2a3e070a6c0c.png">

marcotuts replied:
I didnt find the specific line / class for this but the h6 headings for .section-tools is large, and should be dropped down to match the existing course home. This change involves all sidebar headers, so its likely not scoped to just section tools but I thought I'd note it somewhere

andy-armstrong replied:
Nice catch. That rule lived in a generic Pattern Library partial so it didn't get included in the Bootstrap versions. I moved it into _course-experience.scss.

HarryRein commented on lms/static/sass/bootstrap/_layouts.scss:
```diff
@@ -1,49 +1,81 @@
+
+  .page-header-main {
+    flex-grow: 1;  // This column should consume all the available space
+    min-width: 300px;
```
Oh, wait I see why you have that now. It's leading to the collapse of the button to under the title, we should see if we can find a smoother solution with the bootstrap grid system that at least stretches the "Resume Course" button to full width if it drops under.

andy-armstrong replied:
Agreed.

andy-armstrong commented on common/static/sass/edx-pattern-library-shims/_buttons.scss:
```diff
@@ -86,9 +86,9 @@
 
 // ----------------------------
-// #BRAND
+// #PRIMARY
```
@HarryRein Bootstrap switched from brand over to primary, so I figured we should be doing the same for consistency. For v1 and v2 styles I'm continuing to support brand as an alternative, but eventually I'm thinking we should switch over to `theme-color("primary")` etc.

andy-armstrong commented on lms/static/sass/_build-lms-v1.scss:
```diff
@@ -3,6 +3,12 @@
+@import 'bootstrap/theme';
+@import 'bootstrap/scss/functions';
+@import 'bootstrap/scss/variables';
+@import "bootstrap/scss/mixins/breakpoints";
```
Nice catch. I also did only a quick pass, plus I was spot checking on the sandbox. IMO the Bootstrap colors are the correct ones as they were taken from the Pattern Library. I should probably remove the duplicates so that this is clearer.

andy-armstrong commented on lms/static/sass/bootstrap/_layouts.scss:
```diff
@@ -1,49 +1,81 @@
 
-        .page-content {
-            padding: 20px;
+        @include media-breakpoint-down(md) {
```
I was debating that. There are some instances that have a lot of tabs so this might end up wrapping before the breakpoint. I'll try the smaller value and see how it looks.

andy-armstrong commented on lms/static/sass/bootstrap/_layouts.scss:
```diff
@@ -1,49 +1,81 @@
+  display: flex;
+  flex-wrap: wrap;
+  border-bottom: 1px solid $border-color;
+  padding: 20px;
```
Yeah, let's talk through what you mean. We can certainly evolve any of these settings over the next few weeks, but let me know if there's something concrete you'd like to see in this PR.

AlasdairSwan commented on themes/edx.org/lms/templates/header.html:
```diff
@@ -68,33 +68,33 @@ <h1 class="hd logo-header">
       % endif
     </h1>
-    <button class="navbar-toggler navbar-toggler-right mt-2" type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
+    <button class="navbar-toggler navbar-toggler-right" type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
```
This feature is not accessible. Focus should be trapped in the menu when it is opened.

andy-armstrong replied:
@HarryRein I'll need your help to see what we can do about this. This is just the built-in Bootstrap implementation.

AlasdairSwan commented on themes/edx.org/lms/templates/header.html:
```diff
@@ -68,33 +68,33 @@ <h1 class="hd logo-header">
     </h1>
-    <button class="navbar-toggler navbar-toggler-right mt-2" type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
+    <button class="navbar-toggler navbar-toggler-right" type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
       <span class="navbar-toggler-icon"></span>
```
This should be ```aria-hidden```

andy-armstrong replied:
Fixed

andy-armstrong commented on lms/static/sass/bootstrap/_layouts.scss:
```diff
@@ -1,49 +1,81 @@
-        max-width: map-get($container-max-widths, xl);
+    .page-content-secondary {
+      width: 350px;
+      margin-left: $baseline*2;
```
Yeah, I'd like to talk this through. I don't love adding these classes to every template, especially as the templates are currently shared between v2 and Bootstrap. I prefer to tell folks just to add the one class and that that controls the layout and spacing in a consistent way.

marcotuts replied:
:) This seems like a balancing act between bootstrap's grid class duplication vs layout class styles for pages / templates ... I will flag as a note that the secondary column on course home is narrower than it used to be, so once we decide the approach here we should also make the column  split more even.

andy-armstrong commented on lms/static/sass/features/_course-experience.scss:
```diff
@@ -268,7 +270,7 @@
+      font-size: 0.9rem;
       line-height: 1;
       font-weight: $font-bold;
       color: $lms-gray;
```
Good call. Let me see which is the closest.

marcotuts commented on lms/static/sass/course/layout/_courseware_preview.scss:
```diff
@@ -1,27 +1,31 @@
 
   .preview-actions {
     display: inline-block;
+    margin-bottom: 0;
```
add - @include padding-left(0); to align staff bar
![image](https://user-images.githubusercontent.com/2023680/31292989-8f4dfe24-aaa3-11e7-940e-7fa11bc970f0.png)

andy-armstrong replied:
Fixed

andy-armstrong commented on lms/templates/user_dropdown.html:
```diff
@@ -14,22 +14,24 @@
+                <% profile_image_url = get_profile_image_urls_for_user(self.real_user)['medium'] %>
+                <img class="user-image-frame" src="${profile_image_url}" alt="">
+            </a>
+            <li class="dropdown">
```
Agreed. Fixed.

andy-armstrong commented on lms/templates/user_dropdown.html:
```diff
@@ -14,22 +14,24 @@
+                <% username = self.real_user.username %>
+                <a class="dropdown-toggle" id="navbarDropdownMenuLink" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">${username}</a>
+                <ul class="dropdown-menu dropdown-menu-right" id="${_("Usermenu")}" aria-labelledby="dropdownMenuLink" tabindex="-1">
+                    <a class="dropdown-item" href="${reverse('dashboard')}">${_("Dashboard")}</a>
```
Fixed

andy-armstrong commented on themes/edx.org/lms/templates/header.html:
```diff
@@ -113,13 +113,13 @@ <h1 class="hd logo-header">
           % endif
         % endif
-        <ul class="navbar-nav navbar-right">
+        <ul class="navbar-nav ml-auto">
```
Fixed.

marcotuts commented on lms/static/sass/bootstrap/_layouts.scss:
```diff
@@ -1,49 +1,81 @@
+  }
+
+  &.container-fluid {
+    max-width: map-get($container-max-widths, xl);
```
It looks like we have moved to a max-width here of 1140 instead of 1180. Is that bootstrap's default? If anything, I would encourage us to move to 1200 or 1260?

marcotuts commented:
.date-summaryis missing border-left rules in the course dates sidebar

marcotuts commented:
Open item - Switching course navigation from tabs --> dropdown for smaller screen sizes to collapse the vertical listing of tabs

andy-armstrong commented on lms/static/sass/bootstrap/_layouts.scss:
```diff
@@ -1,49 +1,81 @@
         }
+      }
+
+      .nav-link {
```
Fixed.

AlasdairSwan commented on lms/static/sass/bootstrap/_header.scss:
```diff
@@ -0,0 +1,13 @@
+    margin: 0;
+    padding: 0;
+
+    img {
```
@andy-armstrong why style the element instead of a class?

andy-armstrong replied:
Laziness. :-) There was no class on the image, and it didn't seem necessary to add one. I can change it.

AlasdairSwan replied:
If you could that would be great.

andy-armstrong replied:
Done

AlasdairSwan commented on lms/static/sass/bootstrap/_layouts.scss:
```diff
@@ -1,49 +1,82 @@
-        max-width: map-get($container-max-widths, xl);
+    .page-content-secondary {
+      min-width: 350px;
+      max-width: 350px;
```
Why not just set ```width: 350px```?

andy-armstrong replied:
That's what I had originally, but it was ending up smaller than 350p. Any ideas why that might be?

AlasdairSwan replied:
Was the container set to ```display: flex;```? If so you could try setting ```flex-wrap: wrap'``` on said container.

andy-armstrong replied:
I started out with flex-wrap, but I found that it was wrapping too soon with some browsers. I figured it was better to let the wrapping happen at a breakpoint.

AlasdairSwan replied:
I think it would be worth leaving a comment here explaining this, so if someone in the future figures out how to resolve the issue they can update the code with more confidence.

andy-armstrong replied:
Done

AlasdairSwan commented on lms/static/sass/features/_course-experience.scss:
```diff
@@ -1,11 +1,33 @@
+// Course sidebar sections
+.course-sidebar {
+  .section {
+    h3 {
```
Styles should be applied to classes instead of elements. Especially important for headings given their a11y significance.

andy-armstrong replied:
The challenge here is that arbitrary HTML is rendered by course teams into the course dates section. I'll try adding a class and see if it still works okay.

AlasdairSwan replied:
I remember the nightmare A.J. had adjusting headings for an a11y review, which really hit home the importance of not styling elements directly.

andy-armstrong replied:
I've made this change and it seems fine.

AlasdairSwan replied:
Great, if you can commit that change to this PR that would be good.

andy-armstrong replied:
Done

andy-armstrong commented:
@HarryRein @marcotuts @AlasdairSwan I've fixed the majority of the issues, I updated the sandbox, and then I ran the Axe tool and got no warnings. Please re-review when you have a chance, but bear in mind that we'd like to just merge this to make forward momentum. We'll have more chances to improve a11y and the visual design before this is enabled in production.

AlasdairSwan commented on lms/static/sass/shared-v2/_layouts.scss:
```diff
@@ -1,57 +1,77 @@
-        padding: $baseline;
+      .page-content-secondary {
+        min-width: 350px;
+        max-width: 350px;
```
Why not just ```width: 350px```?

andy-armstrong replied:
Same problem with the width ending up smaller in some instances.

AlasdairSwan replied:
I think it would be worth leaving a comment here explaining this, so if someone in the future figures out how to resolve the issue they can update the code with more confidence.

andy-armstrong replied:
Done

AlasdairSwan commented on lms/templates/user_dropdown.html:
```diff
@@ -14,22 +14,24 @@
+    <div class="nav-item nav-item-hidden-collapsed container">
+        <div class="nav align-items-center">
+            <a class="menu-title mr-2" href="${reverse('dashboard')}">
+                <span class="sr-only">${_("Dashboard for:")}</span>
```
I'm not sure why the username isn't added here. @cptvitamin shouldn't it be?

andy-armstrong replied:
It seems that it is because the username is shown as a link immediately following it.

andy-armstrong replied:
Again this isn't my change, so I'm not sure.

AlasdairSwan replied:
Cool, I'm fine leaving as is.

cptvitamin replied:
@andy-armstrong @AlasdairSwan The username needs to be included.  It used to be included using the alt attribute for the image that follows, but I see that image now has an empty alt attribute (no value).   It looks like this now:
```
<a href="/dashboard" class="user-link">
        <span class="sr">Dashboard for:</span>
        <img class="user-image-frame" src="..." alt="">
        <div class="label-username">username</div>
</a>
``` 
It has to be one or the other, either an alt attr value, or accessible text.

AlasdairSwan commented on themes/edx.org/lms/templates/header.html:
```diff
@@ -104,60 +105,64 @@ <h1 class="hd logo-header">
+          <ul class="nav align-items-center">
+            % if should_display_shopping_cart_func(): # see shoppingcart.context_processor.user_has_cart_context_processor
+              <li>
+                <a role="button" class="nav-item-open-collapsed btn-shopping-cart btn btn-secondary mr-3" href="${reverse('shoppingcart.views.show_cart')}">
```
Why does this have the role of button?

andy-armstrong replied:
I didn't write this code (I think I just indented it). I'll remove the role.

AlasdairSwan commented on themes/red-theme/lms/static/sass/partials/bootstrap/_theme.scss:
```diff
@@ -25,3 +25,6 @@ $theme-colors: map-merge((
   dark: $gray-800
 ), $theme-colors);
+
+// Why is this necessary?
```
Good question, why is it?

andy-armstrong replied:
I was unable to find out why the Bootstrap variable didn't pick up the brand color. I think it is a Sass evaluation order issue, but I couldn't figure it out.

AlasdairSwan replied:
I think it would be worth leaving a comment here explaining this, so if someone in the future figures out how to resolve the issue they can update the code with more confidence.

andy-armstrong replied:
Done

HarryRein commented:
Issue with colors on the course home page pre-start for an un-logged in user:
![screen shot 2017-10-10 at 1 11 00 pm](https://user-images.githubusercontent.com/3861052/31400172-a1b4b9b0-adbc-11e7-91d1-5ac8e4cc0501.png)

Issues on bootstrap page:
[ ] Sidebar appearing on left of main panel
[ ] panel not full width so visual issues with when the main container ends
[ ] Resume button not styling properly
[ ] At medium sized screen sizes, the items collapse correctly but still are styled as if they are open (I think that was in issue with my first review comment, my mistake)
[ ] Course dates styling could use a tiny bit of touch up, I know we will handle it later but at least the header colors

![screen shot 2017-10-10 at 1 13 02 pm](https://user-images.githubusercontent.com/3861052/31400331-321a0b18-adbd-11e7-8f5a-bddc45dfe7b3.png)

HarryRein commented:
Alert icon offset
![screen shot 2017-10-10 at 1 16 35 pm](https://user-images.githubusercontent.com/3861052/31400345-44416dae-adbd-11e7-8edb-27914c26de91.png)

andy-armstrong commented:
@HarryRein I had a bad merge that left the layout and color issues you saw. I'm rebuilding a sandbox with those changes reapplied. I'm tempted to leave some of the minor issues for a subsequent clean up, because most of them are unrelated to my changes.

cptvitamin commented on lms/templates/user_dropdown.html:
```diff
@@ -14,22 +14,24 @@
+            <div class="dropdown">
+                <% username = self.real_user.username %>
+                <a class="dropdown-toggle" id="navbarDropdownMenuLink" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">${username}</a>
+                <ul class="dropdown-menu dropdown-menu-right" id="${_("Usermenu")}" aria-labelledby="dropdownMenuLink" tabindex="-1">
```
ul needs role=menu

cptvitamin commented on lms/templates/user_dropdown.html:
```diff
@@ -14,22 +14,24 @@
+                <% username = self.real_user.username %>
+                <a class="dropdown-toggle" id="navbarDropdownMenuLink" data-toggle="dropdown" aria-haspopup="true" aria-expanded="false">${username}</a>
+                <ul class="dropdown-menu dropdown-menu-right" id="${_("Usermenu")}" aria-labelledby="dropdownMenuLink" tabindex="-1">
+                    <li><a class="dropdown-item" href="${reverse('dashboard')}">${_("Dashboard")}</a></li>
```
```li```  needs role=presentation
```a```  needs role=menuitem

cptvitamin commented on lms/templates/user_dropdown.html:
```diff
@@ -14,22 +14,24 @@
+            </div>
+        </div>
+    </div>
     <li class="nav-item nav-item-open-collapsed-only collapse"><a href="${reverse('dashboard')}">${_("Dashboard")}</a></li>
```
same semantics and roles need to be applied here as well.

andy-armstrong commented:
@cptvitamin Thanks. I've addressed all of your feedback and I'm rebuilding my sandbox now.
